packages/fpdart/lib/src/effect.dart

```diff
@@ -0,0 +1,800 @@
+            Right(value: final value) =>
+              Effect<E, Never, R>.succeed(value)._unsafeRun(context),
+          },
+        ),
+      );
+
+  /// {@category error_handling}
+  Effect<E, C, R> catchCause<C>(
+    Effect<E, C, R> Function(Cause<L> cause) f,
+  ) =>
+      Effect.from(
+        (context) => _unsafeRun(context).then(
+          (exit) => switch (exit) {
+            Left(value: final cause) => f(cause),
+            Right(value: final value) => Effect<E, C, R>.succeed(value),
+          }
+              ._unsafeRun(context),
+        ),
+      );
+
+  /// {@category filtering}
+  Effect<E, L, R> filterOrDie<C>({
+    required bool Function(R r) predicate,
+    required C Function(R r) orDieWith,
+  }) =>
+      Effect.from(
+        (context) => _unsafeRun(context).then(
+          (exit) => switch (exit) {
+            Left(value: final cause) => Left(cause),
+            Right(value: final value) => predicate(value)
+                ? Right(value)
+                : Left(Die.current(orDieWith(value))),
+          },
+        ),
+      );
+
+  /// {@category filtering}
+  Effect<E, L, R> filterOrElse({
+    required bool Function(R r) predicate,
+    required Effect<E, L, R> Function(R r) orElse,
+  }) =>
+      Effect.from(
+        (context) => _unsafeRun(context).then(
+          (exit) => switch (exit) {
+            Left(value: final cause) => Left(cause),
+            Right(value: final value) => predicate(value)
+                ? Right(value)
+                : orElse(value)._unsafeRun(context),
+          },
+        ),
+      );
+
+  /// {@category delay}
+  Effect<E, L, R> delay(Duration duration) =>
+      Effect.sleep<E, L>(duration).zipRight(this);
+
+  /// {@category delay}
+  Effect<E, L, R> timeout(Duration duration) =>
+      race(Effect<E, L, R>.failCause(const Interrupted()).delay(duration));
+
+  /// {@category interruption}
+  Effect<E, Never, R> interrupt() => Effect.failCause(const Interrupted());
+}
+
+extension EffectWithScopeFinalizer<E extends ScopeMixin, L, R>
+    on Effect<E, L, R> {
+  /// {@category scoping}
+  Effect<E, L, R> addFinalizer(Effect<Null, Never, Unit> release) =>
+      tapEnv((_, env) => env.addScopeFinalizer(release));
+
+  /// {@category scoping}
+  Effect<E, L, R> acquireRelease(
+    Effect<Null, Never, Unit> Function(R r) release,
+  ) =>
+      tap((r) => addFinalizer(release(r)));
+}
+
+extension EffectNoScopeFinalizer<E, L, R> on Effect<E, L, R> {
+  /// {@category scoping}
+  Effect<Scope<E>, L, R> get withScope => Effect<Scope<E>, L, R>.from(
+        (ctx) => _unsafeRun(ctx.withEnv(ctx.env.env)),
+      );
+
+  /// {@category scoping}
+  Effect<Scope<E>, L, R> addFinalizer(Effect<Null, Never, Unit> release) =>
+      withScope.tapEnv(
+        (_, env) => env.addScopeFinalizer(release),
+      );
+
+  /// {@category scoping}
+  Effect<Scope<E>, L, R> acquireRelease(
+    Effect<Null, Never, Unit> Function(R r) release,
+  ) =>
+      withScope.tapEnv(
```

Comment:
Technically you should make the resource acquisition uninterruptible here, would require some thought about how to make interruption toggle-able / regional.